frontend/src/lib/k8s/apiProxy/apiTypes.ts

```diff
@@ -0,0 +1,309 @@
+
+export type SingleApiFactoryArguments = [group: string, version: string, resource: string];
+export type MultipleApiFactoryArguments = SingleApiFactoryArguments[];
+export type ApiFactoryArguments = SingleApiFactoryArguments | MultipleApiFactoryArguments;
+
+export type SimpleApiFactoryWithNamespaceArguments = [
+  group: string,
+  version: string,
+  resource: string,
+  includeScale?: boolean
+];
+export type MultipleApiFactoryWithNamespaceArguments = SimpleApiFactoryWithNamespaceArguments[];
+export type ApiFactoryWithNamespaceArguments =
+  | SimpleApiFactoryWithNamespaceArguments
+  | MultipleApiFactoryWithNamespaceArguments;
+
+export interface ApiClient<ResourceType extends KubeObjectInterface> {
+  list: (
+    cb: StreamResultsCb<ResourceType>,
+    errCb: StreamErrCb,
+    queryParams?: QueryParameters,
+    cluster?: string
+  ) => Promise<CancelFunction>;
+  get: (
+    name: string,
+    cb: StreamResultsCb<ResourceType>,
+    errCb: StreamErrCb,
+    queryParams?: QueryParameters,
+    cluster?: string
+  ) => Promise<CancelFunction>;
+  post: (
+    body: RecursivePartial<ResourceType>,
+    queryParams?: QueryParameters,
+    cluster?: string
+  ) => Promise<ResourceType>;
+  put: (
+    body: ResourceType,
+    queryParams?: QueryParameters,
+    cluster?: string
+  ) => Promise<ResourceType>;
+  patch: (
+    body: OpPatch[],
+    name: string,
+    queryParams?: QueryParameters,
+    cluster?: string
+  ) => Promise<ResourceType>;
+  delete: (name: string, queryParams?: QueryParameters, cluster?: string) => Promise<any>;
+  isNamespaced: boolean;
+  apiInfo: {
+    group: string;
+    version: string;
+    resource: string;
+  }[];
+}
+
+export interface ApiWithNamespaceClient<ResourceType extends KubeObjectInterface> {
+  list: (
+    namespace: string,
+    cb: StreamResultsCb<ResourceType>,
+    errCb: StreamErrCb,
+    queryParams?: QueryParameters,
+    cluster?: string
+  ) => Promise<CancelFunction>;
+  get: (
+    namespace: string,
+    name: string,
+    cb: StreamResultsCb<ResourceType>,
+    errCb: StreamErrCb,
+    queryParams?: QueryParameters,
+    cluster?: string
+  ) => Promise<CancelFunction>;
+  post: (
+    body: RecursivePartial<KubeObjectInterface>,
+    queryParams?: QueryParameters,
+    cluster?: string
+  ) => Promise<any>;
+  put: (
+    body: KubeObjectInterface,
+    queryParams?: QueryParameters,
+    cluster?: string
+  ) => Promise<ResourceType>;
+  patch: (
+    body: OpPatch[],
+    namespace: string,
+    name: string,
+    queryParams?: QueryParameters,
+    cluster?: string
+  ) => Promise<any>;
+  delete: (
+    namespace: string,
+    name: string,
+    queryParams?: QueryParameters,
+    cluster?: string
+  ) => Promise<any>;
+  isNamespaced: boolean;
+  apiInfo: {
+    group: string;
+    version: string;
+    resource: string;
+  }[];
+  scale?: ScaleApi;
+}
+
+export interface ScaleApi {
```

Comment:
I wonder if we should break types out into .types.ts files? This is something mui does.

The benefit of having the types next to their implementation is that things are easier to find. scaleApi.types.ts

Reply:
we can consider it. I think current type colocation is good enough